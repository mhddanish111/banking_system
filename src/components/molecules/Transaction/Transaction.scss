@import "../../../styles/colors.scss" ;
@import "../../../styles/spacing.scss";
@import "../../../styles/boarder.scss";
.transaction-container {
  margin-top: $space-l;
  .transaction-table {
    width: 100%;
    border-collapse: collapse;
    overflow-x: scroll;
    margin-top: $space-s;
    border: $border-dark;
    thead {
      border: $border-dark;
    }
    th,
    td {
      text-align: left;
      padding: 8px;
    }
    tr:nth-child(even) {
      background-color: $color-white;
    }
  }
  .transaction-table-container {
    overflow-x: auto;
  }
  .pagination {
    padding: 0;
    display: flex;
    justify-content: flex-end;
  }
}

