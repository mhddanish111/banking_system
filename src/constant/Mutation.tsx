import { gql } from '@apollo/client';

export const CREATE_USER = gql`
  mutation CreateUser($userInput: UserInput) {
    createUser(userInput: $userInput) {
      firstName
      lastName
      message
    }
  }
`;

export const CHANGE_PASSWORD = gql`
  mutation ChangePassword($changePasswordInput: ChangePasswordInput) {
    changePassword(changePasswordInput: $changePasswordInput) {
      message
      success
      data {
        firstName
      }
    }
  }
`;

export const OPEN_ACCOUNT = gql`
  mutation CreateAccount($accountInput: AccountInput) {
    createAccount(accountInput: $accountInput) {
      _id
      accountNumber
      amount
      accountType
      categoryName
      limit
      minAmount
      maxAmount
      accountTypeId
      categoryId
      userId
      message
      type
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
    }
  }
`;

export const DEPOSIT = gql`
  mutation Deposit($depositInput: DepositInput) {
    deposit(depositInput: $depositInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;

export const WITHDRAWAL = gql`
  mutation Withdrawal($withdrawalInput: WithdrawalInput) {
    withdrawal(withdrawalInput: $withdrawalInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;
