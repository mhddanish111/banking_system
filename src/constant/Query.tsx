import { gql } from '@apollo/client';

export const LOGIN_USER = gql`
  query Login($logInInput: LogInInput) {
    login(logInInput: $logInInput) {
      firstName
      lastName
      userId
      token
      isValid
      message
      userType
    }
  }
`;

export const USER_DETAILS = gql`
  query UserDetails {
    userDetails {
      _id
      firstName
      lastName
      email
      mobile
      occupation
      income
      aadhaarNumber
      panNumber
      govtId
      govtIdNumber
      address
      city
      state
      postalCode
      isActive
      message
    }
  }
`;

export const GET_ACCOUNT_TYPES = gql`
  query AccountTypes {
    accountType {
      _id
      name
      type
      isActive
      category {
        _id
        name
        benefit
        isActive
        minAmount
        maxAmount
        limit
        charge
      }
    }
  }
`;

export const GET_ACCOUNTS = gql`
  query Accounts {
    accounts {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
      }
    }
  }
`;

export const GET_ACCOUNT_WITH_TYPE = gql`
  query AccountWithType($accountWithTypeInput: AccountWithTypeInput) {
    accountWithType(accountWithTypeInput: $accountWithTypeInput) {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
        aadhaarNumber
        panNumber
      }
    }
  }
`;

export const GET_TRANSACTION = gql`
  query Transaction($transactionInput: TransactionInput) {
    transaction(transactionInput: $transactionInput) {
      count {
        offset
        limit
        totalCount
      }
      result {
        _id
        transactionId
        accountNumber
        amount
        withdrawalAmount
        depositAmount
        transactionDate
        accountType
        categoryName
        transactionType
      }
    }
  }
`;
