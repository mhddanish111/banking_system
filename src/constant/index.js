export const USER_KEY = 'user_key';

export const AccountTypeList = [
  { label: 'Accounts', value: 'Accounts' },
  { label: 'Cards', value: 'Cards' },
];
