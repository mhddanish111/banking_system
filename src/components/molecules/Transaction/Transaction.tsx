import React, { useState, useEffect, useCallback } from 'react';
import { useLazyQuery } from '@apollo/client';
import moment from 'moment';
import Heading from '../../atoms/Heading';
import { GET_TRANSACTION } from '../../../constant/Query';
import Pagination from '../../atoms/Pagination';
import TransactionSearch from '../TransactionSearch';
import './Transaction.scss';

type SearchProps = {
  fromDate: string;
  toDate: string;
};
const Transaction: React.FC<any> = (props) => {
  const {
    selectedAccountNumber,
    setIsCallTransaction,
    isCallTransaction = false,
  } = props;
  const limit = 10;
  const [getTransaction] = useLazyQuery(GET_TRANSACTION);
  const [isOnPageLoad, setIsPageOnLoad] = useState<boolean>(true);
  const [offset, setOffset] = useState(1);
  const [search, setSearch] = useState({
    showAll: true,
    last10Transaction: false,
    customSearch: false,
  });
  // const [isCustomSearch, setCustomSearch] = useState<boolean>(false);
  const [transaction, setTransaction] = useState({
    count: {
      offset: 0,
      limit,
      totalCount: 10,
    },
    result: [],
  });
  const callTransaction = (page: any) => {
    setOffset(page);
    setIsPageOnLoad(true);
  };

  const handleTransaction = useCallback(
    (props: any) => {
      getTransaction({
        variables: { transactionInput: props },
        fetchPolicy: 'network-only',
      })
        .then((res) => {
          const transaction = res?.data?.transaction;
          setTransaction(transaction);
          setIsPageOnLoad(false);
          setIsCallTransaction(false);
        })
        .catch((err) => {
          setIsPageOnLoad(false);
          setIsCallTransaction(false);
        });
    },
    [getTransaction, setIsCallTransaction],
  );
  useEffect(() => {
    if ((isOnPageLoad || isCallTransaction) && selectedAccountNumber) {
      const { showAll, last10Transaction, customSearch } = search;
      const date = new Date().toISOString();
      const props = {
        accountNumber: selectedAccountNumber,
        offset: offset - 1,
        limit,
        fromDate: date,
        toDate: date,
        showAll,
        last10Transaction,
        customSearch,
      };
      handleTransaction(props);
    }
  }, [
    offset,
    selectedAccountNumber,
    isOnPageLoad,
    isCallTransaction,
    setIsPageOnLoad,
    getTransaction,
    setIsCallTransaction,
    handleTransaction,
    search,
  ]);
  const searchHandler = (data: SearchProps) => {
    const date = new Date().toISOString();
    const { showAll, last10Transaction, customSearch } = search;
    const { fromDate = date, toDate = date } = data;
    const props = {
      accountNumber: selectedAccountNumber,
      offset: offset - 1,
      limit,
      fromDate,
      toDate,
      showAll,
      last10Transaction,
      customSearch,
    };
    handleTransaction(props);
  };
  return (
    <div className="row transaction-container">
      {transaction?.result?.length > 0 && (
        <>
          <div className="col-xs-12">
            <Heading HeadingType="h3">Transaction Details</Heading>
          </div>
          <div className="col-xs-12">
            <TransactionSearch
              searchHandler={searchHandler}
              search={search}
              setSearch={setSearch}
            />
          </div>
          <div className="col-sm-12 col-xs-12 pagination">
            <Pagination
              className="pagination-bar"
              currentPage={offset}
              totalCount={transaction?.count?.totalCount}
              pageSize={limit}
              onPageChange={(page: any) => callTransaction(page)}
            />
          </div>
          <div className="col-xs-12 transaction-table-container">
            <table className="transaction-table">
              <thead>
                <tr>
                  <th>S.No.</th>
                  <th>Transaction Id</th>
                  <th>Transaction Date</th>
                  <th>Transaction Type</th>
                  <th>Withdrawal</th>
                  <th>Deposit</th>
                  <th>Balance</th>
                </tr>
              </thead>
              <tbody>
                {transaction.result.map((item, index) => {
                  const {
                    transactionId,
                    transactionDate,
                    transactionType,
                    withdrawalAmount,
                    depositAmount,
                    amount,
                  } = item;
                  const date = moment(transactionDate).format('DD/MM/YYYY');
                  return (
                    <tr key={transactionId}>
                      <td>{index + 1}</td>
                      <td>{transactionId}</td>
                      <td>{date}</td>
                      <td>{transactionType}</td>
                      <td>{withdrawalAmount}</td>
                      <td>{depositAmount}</td>
                      <td>{amount}</td>
                    </tr>
                  );
                })}
              </tbody>
            </table>
          </div>
        </>
      )}
    </div>
  );
};

export default Transaction;
