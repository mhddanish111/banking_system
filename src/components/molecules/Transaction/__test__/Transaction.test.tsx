import '@testing-library/jest-dom';
import { MockedProvider } from '@apollo/client/testing';
import renderer from 'react-test-renderer';
import Transaction from '../Transaction';
import { GET_TRANSACTION } from '../../../../constant/Query';

describe('<Transaction />', () => {
  const props = {
    selectedAccountNumber: 1673894730214,
    setIsCallTransaction: jest.fn(),
    isCallTransaction: true,
  };
  const submitProps = {
    accountNumber: 1234567890,
    payeeAccountNumber: 12123123123123,
    amount: 100,
    accountType: 'Accounts',
  };
  const mocks = [
    {
      request: {
        query: GET_TRANSACTION,
        variables: {
          depositInput: submitProps,
        },
      },
      result: {
        data: {
          transaction: {
            _id: '12345',
          },
        },
      },
    },
  ];
  const wrapper = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Transaction {...props} />
    </MockedProvider>,
  ).toJSON;
  it('should render the Transaction Component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
