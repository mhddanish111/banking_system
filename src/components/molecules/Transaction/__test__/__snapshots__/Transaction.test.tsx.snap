// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transaction /> should render the Transaction Component correctly 1`] = `[Function]`;
